components/pango_context/include/pangolin/gui/layer_group.h

```diff
@@ -23,22 +25,52 @@ struct LayerGroup {
                  // available space. Requires common aspect for each layer.
   };
 
-  LayerGroup() = default;
-  LayerGroup(Shared<Layer> layer) : layer(layer) {}
-
-  Grouping grouping = Grouping::horizontal;
-  std::vector<LayerGroup> children = {};
-  std::shared_ptr<Layer> layer = nullptr;
-  size_t selected_tab = 0;
+  struct Params {
+    Grouping grouping = Grouping::horizontal;
+    std::shared_ptr<Layer> layer = nullptr;
+    size_t selected_tab = 0;
+  };
 
-  struct LayoutInfo {
+  struct Constraints {
+    // todo: Describe semantic: What is the precise definition of min_pix?  What
+    // is the precise definition of parts?
     Eigen::Array2i min_pix = {0, 0};
     Eigen::Array2d parts = {0.0f, 0.0f};
-    double aspect_hint = 0.0;
-    Region2I region = Region2I::empty();
+    // Hint / recommendation for what aspect ratio to use.
+    AspectRatio aspect = {};
   };
 
-  mutable LayoutInfo cached_;
+  LayerGroup() = default;
+  LayerGroup(const Params& params) : params_(params) {}
+
+  const Params& params() const { return params_; }
+
+  bool hasConstraint() const { return bool(constraints_); }
+
+  void setConstraints(const Constraints& constraints)
+  {
+    constraints_ = constraints;
+  }
+
+  // Precondition: this->hasConstraint() must be true
+  const Constraints& constraints() const { return FARM_UNWRAP(constraints_); }
```

Comment:
I think it's best to just return a const ref to the optional instead of having a hasConstraints() and constraints() method.

Reply:
good idea will done